text/0000-compiler-plugins.md

```diff
@@ -0,0 +1,148 @@
+* Start Date: 2018-07-24
+* RFC PR: (leave this empty)
+* Fusion Issue: (leave this empty)
+
+# Summary
+
+Fusion plugins should be able to perform operations at build-time.
+
+# Basic example
+
+```js
+// src/app.js
+
+import {foo} from "fusion-plugin-foo";
+
+foo();
+```
+
+```json
+{
+  "name": "fusion-plugin-foo",
+  "...": "...",
+  "fusion": {
+    "compiler-plugin": "./compiler-plugin.js"
+  }
+}
+```
+
+```js
+// node_modules/fusion-plugin-foo/compiler-plugin.js
+module.exports = function plugin() {
+  return {
+    analyzer: (...babelPluginArgs) => {
+      // Traverse AST to count number of times `foo` is invoked
+      return {callCount};
+    },
+    transformer: (...babelPluginArgs) => {
+      // Conceptually equivalent to "babel-macros" idea
+      // Statically replace `foo()` CallExpression with "foo" StringLiteral
+    },
+    aggregator: (allMetadata, compiler) => {
```

Comment:
I’ve been thinking about an alternative to returning JS source code. Instead, the aggregator could return a serializable value that gets registered. I suppose in this case, the compiler plugin must provide a token (probably in the form of a path to a file that exports the token).

Then the package containing the compiler plugin could easily consume the aggregated values via the DI system.

Reply:
Some nice properties of using the DI system:

1. If a buildtime token is not registered (not using fusion-CLI, for example), there’s an error.
2. Since buildtime values can be easily transformed or mocked using their associated token.